components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -0,0 +1,162 @@
+import kotlinx.coroutines.future.await
+import kotlinx.coroutines.guava.asDeferred
+import org.jetbrains.plugins.terminal.TerminalToolWindowFactory
+import org.jetbrains.plugins.terminal.TerminalView
+
+@Suppress("UnstableApiUsage", "EXPERIMENTAL_IS_NOT_ENABLED", "OPT_IN_IS_NOT_ENABLED")
+@OptIn(DelicateCoroutinesApi::class)
+class GitpodTerminalService(private val project: Project) : Disposable {
+    companion object {
+        val TITLE_KEY = Key.create<String>("TITLE_KEY")
+    }
+
+    private val lifetime = Lifetime.Eternal.createNested()
+    private val terminalView = TerminalView.getInstance(project)
+    private val terminalServiceFutureStub =
+            TerminalServiceGrpc.newFutureStub(GitpodManager.supervisorChannel)
+    private val statusServiceStub = StatusServiceGrpc.newStub(GitpodManager.supervisorChannel)
+    private val backendTerminalManager = BackendTerminalManager.getInstance(project)
+
+    override fun dispose() {
+        lifetime.terminate()
+    }
+
+    init {
+        GlobalScope.launch {
+            getTerminalToolWindowRegisteredEvent().await()
+            val tasksList = getSupervisorTasksList().await()
+            val terminalsList = getSupervisorTerminalsListAsync().await().terminalsList
+            debug("Got a list of Supervisor terminals: $terminalsList")
+            runInEdt {
+                if (tasksList.isEmpty()) {
+                    backendTerminalManager.createNewSharedTerminal("Gitpod", "Terminal")
+                } else {
+                    val aliasToTerminalMap: MutableMap<String, TerminalOuterClass.Terminal> =
+                            mutableMapOf()
+
+                    for (terminal in terminalsList) {
+                        val terminalAlias = terminal.alias
+                        aliasToTerminalMap[terminalAlias] = terminal
+                    }
+
+                    for (task in tasksList) {
+                        val terminalAlias = task.terminal
+                        val terminal = aliasToTerminalMap[terminalAlias]
+
+                        if (terminal != null) {
+                            createSharedTerminal(terminal)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun getTerminalToolWindowRegisteredEvent(): CompletableFuture<Void> {
+        val completableFuture = CompletableFuture<Void>()
+
+        debug("Waiting for TerminalToolWindow to be registered...")
+        val toolWindowManagerListener =
+                object : ToolWindowManagerListener {
+                    override fun toolWindowsRegistered(
+                            ids: MutableList<String>,
+                            toolWindowManager: ToolWindowManager
+                    ) {
+                        if (ids.contains(TerminalToolWindowFactory.TOOL_WINDOW_ID)) {
+                            debug("TerminalToolWindow got registered!")
+                            completableFuture.complete(null)
```

Comment:
unsubscribe here?

Reply:
Should be unsubscribed now. (Ref: https://github.com/gitpod-io/gitpod/pull/8338/commits/77e9d0e71208d77ce61a9d1b2e66c50740600473#diff-5e11fb1aae2ce12b8efdde7b6b5e097f14940fd2bc30a775f58c48241f18ed73R94)

The only way I found to unsubscribe was by disposing the project's message bus connection.